plutus-core/plutus-core/src/PlutusCore/Parser/ParserCommon.hs

```diff
@@ -21,9 +21,19 @@ import PlutusCore.Error
 import PlutusCore.Name
 import PlutusCore.Quote
 
+{- Note [Whitespace invariant]
+Every top-level 'Parser' must consume all whitespace after the thing that it parses, hence make
+sure to enclose every 'Parser' that doesn't consume trailing whitespce (e.g. 'takeWhileP',
+'manyTill', 'Lex.decimal' etc) in a call to 'lexeme'.
+-}
+
 topSourcePos :: SourcePos
 topSourcePos = initialPos "top"
 
+-- | The opposite of @optional :: Alternative f => f a -> f (Maybe a)@.
```

Comment:
Then why the `MonadPlus` constraint instead of `Alternative`?

Reply:
Don't see how it could be defined with only `Alternative`. It's "morally" the opposite, not literally.